Sources/SwiftSyntax/BumpPtrAllocator.swift

```diff
@@ -113,8 +111,9 @@ public class BumpPtrAllocator {
       return allocated
     }
 
-    // If the size is too big, allocate a dedicated slab for it.
-    if byteCount >= self.slabSize {
+    // If the size is too large to be contained in the new slab,
+    // allocate a dedicated slab for it.
+    if byteCount > self.slabSize(at: slabs.count) {
```

Comment:
I'm not sure about this, the current semantics of "custom size slab" for _reasonably_ big allocations.

For example, say the init slab size is `1024`, the current slab size is `4096`, and we already used `128` bytes of the current slab. If the next allocation requested `4000` bytes, the current slab cannot hold it, the next can, but if we do that, we'd waste `3968` bytes of the current slab. Custom slab avoid that kind of waste.

Reply:
That makes sense. 
But in the case of your example, if the allocator continues to be requested `4000` bytes, a memory allocation occurs as per the request, and I think this should be avoided. I'm not sure if that could be problematic in our use cases because I'm quite new to this project. If so, I suppose we need a more sophisticated mechanism to ensure that we minimize memory waste and avoid frequent memory allocations.
If that's not be problem in our use cases, do you think it would be better to clarify that we do our best not to waste memory here? If so, I'll add some comments.
@rintaro

Reply:
Each object allocation SwiftSyntax is below 64 bytes. Layout/collection data can be long but it's just (pointer size * number-of-elements), so several hundreds won't be an issue in ParsingArena (slabSize == 4096). Another one is the source buffer which totally makes sense to use a custom size slab. For non-parsing arena (sizeSize == 128) which usually used for editing, memory allocation is not a dominant factor in performance, afaik.
We also use BumpPtrAllocator in ASTGen, but it's similar.

We could make an option to change the behavior, but I don't think it's worth it at this point.

(FWIW, the current behavior resembles [the default behavior of `llvm::BumpPtrAllocator`'s `SizeThreshold` which defaults to `SlabSize`](https://github.com/apple/llvm-project/blob/1fbdb6c1d020d27c18546f422c7b7f5c781aed4c/llvm/include/llvm/Support/Allocator.h#L61-L62) )